### PR TITLE
feat(core): allow removal of previously registered DestroyRef callbacks

### PR DESCRIPTION
**Pull request by pkozlowski-opensource**
This change makes it possible to remove a previously registered destroy callback - to do so it is enough to call the unregistration function returned from the onDestroy method call.

### CODE REVIEW

**Review by alxhub — APPROVED** [review]

**Review by atscott — APPROVED** [review]
reviewed-for: public-api, size-tracking

**Review by AndrewKushnir — APPROVED** [review]
LGTM 👍 Just left a few minor comments, but **none** of them are blocking.

**Review by jessicajaniuk — APPROVED** [review]
reviewed-for: public-api

File: packages/core/src/di/r3_injector.ts [hunk 1]
```diff
@@ -215,6 +220,13 @@ export class R3Injector extends EnvironmentInjector {
     this._onDestroyHooks.push(callback);
   }
 
+  override removeOnDestroy(callback: () => void): void {
```
[the remark this answers is not in the document]
Comment (alxhub) [reply packages/core/src/di/r3_injector.ts:223]: +1, I think this would clean things up considerably

**Review by alxhub — APPROVED** [review]
Reviewed-for: public-api

File: packages/core/src/di/r3_injector.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (pkozlowski-opensource) [reply packages/core/src/di/r3_injector.ts:223]: Yeh, was considering it but I wasn't sure about the need for the un-registration function in other, non-`DestroyRef` cases. 

I don't mind either way.

File: packages/core/src/linker/destroy_ref.ts [hunk 2]
```diff
@@ -38,16 +39,30 @@ export abstract class DestroyRef {
...
-  static __NG_ENV_ID__: (injector: EnvironmentInjector) => DestroyRef = (injector) => injector;
+  static __NG_ENV_ID__: (injector: EnvironmentInjector) => DestroyRef = (injector) => {
+    return new EnvInjectorDestroyRef(injector);
+  }
+}
+
+class EnvInjectorDestroyRef extends DestroyRef {
+  constructor(private _injector: EnvironmentInjector) {
+    super();
+  }
```
[the remark this answers is not in the document]
Comment (pkozlowski-opensource) [reply packages/core/src/linker/destroy_ref.ts:50]: I don't think we can do it here as we are probably not in the injection context. But if we go with the idea of the `EnvInjectorDestroyRef` removal than it is no longer relevant / needed.

File: packages/core/src/linker/destroy_ref.ts [hunk 3]
```diff
@@ -24,9 +24,10 @@ export abstract class DestroyRef {
   // - `EnvironmentInjector` when retrieved from an environment injector
 
   /**
-   * Registers a destroy callback in a given lifecycle scope.
+   * Registers a destroy callback in a given lifecycle scope. A previously registered callback can
+   * be removed by calling the returned unreregistration function.
```
[the remark this answers is not in the document]
Comment (pkozlowski-opensource) [reply packages/core/src/linker/destroy_ref.ts:28]: Done, thnx!

File: packages/core/test/acceptance/destroy_ref_spec.ts [hunk 4]
```diff
@@ -22,6 +22,19 @@ describe('DestroyRef', () => {
       envInjector.destroy();
       expect(destroyed).toBe(true);
     });
+
+    it('should allow removal of destroy callbacks', () => {
```
[the remark this answers is not in the document]
Comment (pkozlowski-opensource) [reply packages/core/test/acceptance/destroy_ref_spec.ts:26]: Yep, I was going back and forth between unregistering the "first one" vs. unregistering all of them... In any case I agree that it is good to add a test that illustrates the current behaviour.

**Pull request closed without merging**